kroxylicious-filters/kroxylicious-oauthbearer-validation/src/main/java/io/kroxylicious/proxy/filter/oauthbearer/OauthBearerValidationFilter.java

```diff
@@ -125,23 +131,31 @@ public CompletionStage<RequestFilterResult> onSaslAuthenticateRequest(short apiV
                         .setErrorMessage("Unexpected SASL request")
                         .setAuthBytes(request.authBytes());
                 LOGGER.debug("SASL invalid state");
+                context.clientSaslAuthenticationFailure(OAUTHBEARER_MECHANISM, null, INVALID_SASL_STATE_EXCEPTION);
                 return context.requestFilterResultBuilder().shortCircuitResponse(failedResponse).withCloseConnection().completed();
             }
             this.saslServer = null;
 
             return authenticate(server, request.authBytes())
                     .thenCompose(bytes -> context.forwardRequest(header, request))
                     .exceptionallyCompose(e -> {
-                        if (e.getCause() instanceof SaslAuthenticationException) {
+                        if (e.getCause() instanceof SaslAuthenticationException cause) {
                             SaslAuthenticateResponseData failedResponse = new SaslAuthenticateResponseData()
                                     .setErrorCode(SASL_AUTHENTICATION_FAILED.code())
                                     .setErrorMessage(e.getMessage())
                                     .setAuthBytes(request.authBytes());
```

Comment:
Same comment, its not your code (and I should admit that I've not read the RFC for OAUTHBEARER mechanism yet), but conceptually it's strange to be responding with the _client's_ `authBytes`. It would make more sense to be using the `bytes` variable here, wouldn't it?